node/src/components/chain_synchronizer/operations.rs

```diff
@@ -948,34 +954,149 @@ where
     block_hash: BlockHash,
     ctx: &ChainSyncContext<'_, REv>,
-) -> Result<Box<BlockAndDeploys>, FetcherError<BlockAndDeploys>>
+) -> Result<Box<BlockAndDeploys>, FetchAndStoreBlockError>
 where
-    REv: From<NetworkInfoRequest> + From<FetcherRequest<BlockAndDeploys>> + From<StorageRequest>,
+    REv: From<NetworkInfoRequest>
+        + From<FetcherRequest<BlockAndDeploys>>
+        + From<StorageRequest>
+        + From<ContractRuntimeRequest>
+        + From<BlocklistAnnouncement>,
 {
     let start = Timestamp::now();
-    let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash).await?;
-    let res = match fetched_block {
-        FetchedData::FromStorage {
-            item: block_and_deploys,
-            ..
-        } => Ok(block_and_deploys),
-        FetchedData::FromPeer {
-            item: block_and_deploys,
-            ..
-        } => {
-            ctx.effect_builder
-                .put_block_and_deploys_to_storage(block_and_deploys.clone())
-                .await;
-            Ok(block_and_deploys)
-        }
-    };
-    ctx.metrics
-        .observe_fetch_block_and_deploys_duration_seconds(start);
-    res
+    'a: loop {
+        let fetched_block = fetch_retry_forever::<_, BlockAndDeploys>(ctx, block_hash)
+            .await
+            .map_err(FetchAndStoreBlockError::Fetcher)?;
+        let res = match fetched_block {
+            FetchedData::FromStorage {
+                item: block_and_deploys,
+                ..
+            } => Ok(block_and_deploys),
+            FetchedData::FromPeer {
+                item: block_and_deploys,
+                peer,
+            } => {
+                if let Some(deploy_approvals_root_hash) =
+                    fetch_deploy_approvals_root_hash(block_and_deploys.block.height(), ctx).await?
+                {
+                    let approvals: BTreeMap<DeployHash, BTreeSet<Approval>> = block_and_deploys
+                        .deploys
+                        .iter()
+                        .map(|d| (*d.id(), d.approvals().clone()))
+                        .collect();
+                    let mut approval_hashes = vec![];
+                    for (dh, maybe_approvals) in block_and_deploys
+                        .block
+                        .body()
+                        .deploy_hashes()
+                        .iter()
+                        .chain(block_and_deploys.block.body().transfer_hashes().iter())
+                        .map(|dh| (*dh, approvals.get(dh)))
+                    {
+                        if let Some(approvals) = maybe_approvals {
+                            approval_hashes.push(Digest::hash(
+                                approvals
+                                    .to_bytes()
+                                    .map_err(FetchAndStoreBlockError::BytesRepr)?,
+                            ));
+                        } else {
+                            error!("no approvals for deploy {:?}", dh);
```

Comment:
```suggestion
                            error!(deploy_hash=?dh, "no approvals for deploy");
```
That way we will get an error log in a JSON format where `deploy_hash` is a field in the JSON object, rather than it being a part of a log message, which can be used for querying.

Current:
```json 
{"message": "no approvals for deploy af1c3..."}
```
suggested:
```json
{"message": "no approvals for deploy", "deploy_hash": "af1c3..."}
```